### ISSUE TITLE
Stop when videos are exhausted?

### ISSUE THREAD

**Issue by SiftingSands** [NVIDIA/DALI#2760]
Referring to the tutorial where multiple videos are read from a directory : https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/sequence_processing/video/video_reader_label_example.html

The pipeline is iterated by a user specified number. However, is there a way to stop the process when all the videos are read to completion? Right now, I believe the behavior is to restart from the beginning if random_shuffle is False. 

My main application in mind is inference on a set of videos. I've got some hacky methods in-mind, but I wanted to check if there was built-in functionality. I've seen some previous issues about using the ExternalSource pipeline, but not sure if that's a good solution here.

Thanks!

**Comment by JanuszL** [comment]
Hi @SiftingSands,

DALI readers are infinite data sources and don't have a very limited notion of the epoch.
For that purpose DALI FW iterators exists, like this one for [Pytorch](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/plugins/pytorch_plugin_api.html). So should do in your case:
```
fw_iter = DALIGenericIterator(pipe, ['data', 'label'], reader_name="reader")
for data in fw_iter:
   # do something with data
```
You can always learn the number of samples that your reader sees `pipe.reader_meta("Reader")["epoch_size"]`. If you devise it by the batch size and number of shards (if applicable) then you should get the number of iterations needed to traverse the whole dataset.

**Comment by SiftingSands** [comment]
Hi @JanuszL 

Wrapping it around a iterator works, but I'm stuck getting the last partial batch. When I set the stride to 1, I get timestamps corresponding to the end of the video at the last iteration. However, when the stride is larger, say 20, then I'm not receiving a partial batch., even though I set that policy in the iterator (following https://docs.nvidia.com/deeplearning/dali/user-guide/docs/plugins/pytorch_plugin_api.html?highlight=base_iterator%20lastbatchpolicy).

Time with stride=1:
```
tensor([[183.4667, 183.5333, 183.6000, 183.6667, 183.7333, 183.8000, 183.8667,
         183.9333, 184.0000, 184.0667, 184.1333, 184.2000, 184.2667, 184.3333,
         184.4000, 184.4667, 184.5333, 184.6000, 184.6667, 184.7333, 184.8000,
         184.8667, 184.9333, 185.0000, 185.0667, 185.1333, 185.2000, 185.2667,
         185.3333, 185.4000, 185.4667, 185.5333, 185.6000, 185.6667, 185.7333,
         185.8000, 185.8667, 185.9333, 186.0000, 186.0667, 186.1333, 186.2000,
         186.2667, 186.3333, 186.4000, 186.4667, 186.5333, 186.6000, 186.6667,
         186.7333, 186.8000, 186.8667, 186.9333, 187.0000, 187.0667, 187.1333,
         187.2000, 187.2667, 187.3333, 187.4000, 187.4667, 187.5333, 187.6000,
         187.6667]], device='cuda:0', dtype=torch.float64)

```

Time with stride=20:
```
tensor([[ 85.3333,  86.6667,  88.0000,  89.3333,  90.6667,  92.0000,  93.3333,
          94.6667,  96.0000,  97.3333,  98.6667, 100.0000, 101.3333, 102.6667,
         104.0000, 105.3333, 106.6667, 108.0000, 109.3333, 110.6667, 112.0000,
         113.3333, 114.6667, 116.0000, 117.3333, 118.6667, 120.0000, 121.3333,
         122.6667, 124.0000, 125.3333, 126.6667, 128.0000, 129.3333, 130.6667,
         132.0000, 133.3333, 134.6667, 136.0000, 137.3333, 138.6667, 140.0000,
         141.3333, 142.6667, 144.0000, 145.3333, 146.6667, 148.0000, 149.3333,
         150.6667, 152.0000, 153.3333, 154.6667, 156.0000, 157.3333, 158.6667,
         160.0000, 161.3333, 162.6667, 164.0000, 165.3333, 166.6667, 168.0000,
         169.3333]], device='cuda:0', dtype=torch.float64)
```

Code snippet for my pipeline - mainly following https://github.com/NVIDIA/DALI/blob/master/docs/examples/use_cases/video_superres/dataloading/dataloaders.py
```
from nvidia.dali.pipeline import Pipeline
from nvidia.dali.plugin import pytorch
import nvidia.dali.fn as fn
import nvidia.dali.types as types

def create_video_reader_pipeline(batch_size, sequence_length, num_threads, 
                                 device_id, files, labels):
    pipeline = Pipeline(batch_size, num_threads, device_id, seed=12)
    with pipeline:
        images, label, start_frame_num, timestamps = fn.video_reader(device="gpu", 
                                 filenames=files, 
                                 sequence_length=sequence_length,
                                 normalized=False, random_shuffle=False, 
                                 image_type=types.RGB, 
                                 dtype=types.UINT8, 
                                 enable_frame_num=True,
                                 enable_timestamps=True,
                                 labels=labels, 
                                 stride=frame_stride,
                                 pad_last_batch=False, name="Reader")

        pipeline.set_outputs(images, label, start_frame_num, timestamps)
    return pipeline

class DALILoader():
    def __init__(self, batch_size, video_filenames, sequence_length, frame_stride,
                 labels, num_threads, device_id):
        self.pipeline = create_video_reader_pipeline(batch_size, 
                                                     sequence_length, 
                                                     num_threads, 
                                                     device_id, 
                                                     video_filenames, 
                                                     labels)
        self.pipeline.build()
        self.epoch_size = self.pipeline.epoch_size("Reader")
        
        output_names = ["data", 'label', 'start_frame_num', 'timestamps']
        self.dali_iterator = pytorch.DALIGenericIterator(self.pipeline,
                                                         output_names,
                                                         reader_name="Reader",
                                                         last_batch_policy=pytorch.LastBatchPolicy.PARTIAL,
                                                         auto_reset=False,
                                                         dynamic_shape=True)

    def __len__(self):
        return int(self.epoch_size)

    def __iter__(self):
        return self.dali_iterator.__iter__()
```

I'm probably missing something incredibly obvious here.

**Comment by JanuszL** [comment]
@SiftingSands ,

> However, when the stride is larger, say 20, then I'm not receiving a partial batch, even though I set that policy in the iterator

Would you normally get this sequence without the DALIGenericIterator?
What is worth noticing is DALI returns only complete sequences for a given stride, step, and length. So while it is possible to build a sequence of given length and stride 1, starting from 187.6667, it is not possible to do so when you have a sequence with a stride of 20.

**Comment by SiftingSands** [comment]
Without DALIGenericIterator, I see the same behavior where no partial batches. It restarts after 169.3 without giving a partial.

```
[[ 0.         1.3333333  2.6666667  4.         5.3333333  6.6666667
   8.         9.3333333 10.6666667 12.        13.3333333 14.6666667
  16.        17.3333333 18.6666667 20.        21.3333333 22.6666667
  24.        25.3333333 26.6666667 28.        29.3333333 30.6666667
  32.        33.3333333 34.6666667 36.        37.3333333 38.6666667
  40.        41.3333333 42.6666667 44.        45.3333333 46.6666667
  48.        49.3333333 50.6666667 52.        53.3333333 54.6666667
  56.        57.3333333 58.6666667 60.        61.3333333 62.6666667
  64.        65.3333333 66.6666667 68.        69.3333333 70.6666667
  72.        73.3333333 74.6666667 76.        77.3333333 78.6666667
  80.        81.3333333 82.6666667 84.       ]]
[[ 85.3333333  86.6666667  88.         89.3333333  90.6666667  92.
   93.3333333  94.6666667  96.         97.3333333  98.6666667 100.
  101.3333333 102.6666667 104.        105.3333333 106.6666667 108.
  109.3333333 110.6666667 112.        113.3333333 114.6666667 116.
  117.3333333 118.6666667 120.        121.3333333 122.6666667 124.
  125.3333333 126.6666667 128.        129.3333333 130.6666667 132.
  133.3333333 134.6666667 136.        137.3333333 138.6666667 140.
  141.3333333 142.6666667 144.        145.3333333 146.6666667 148.
  149.3333333 150.6666667 152.        153.3333333 154.6666667 156.
  157.3333333 158.6666667 160.        161.3333333 162.6666667 164.
  165.3333333 166.6666667 168.        169.3333333]]
[[ 0.         1.3333333  2.6666667  4.         5.3333333  6.6666667
   8.         9.3333333 10.6666667 12.        13.3333333 14.6666667
  16.        17.3333333 18.6666667 20.        21.3333333 22.6666667
  24.        25.3333333 26.6666667 28.        29.3333333 30.6666667
  32.        33.3333333 34.6666667 36.        37.3333333 38.6666667
  40.        41.3333333 42.6666667 44.        45.3333333 46.6666667
  48.        49.3333333 50.6666667 52.        53.3333333 54.6666667
  56.        57.3333333 58.6666667 60.        61.3333333 62.6666667
  64.        65.3333333 66.6666667 68.        69.3333333 70.6666667
  72.        73.3333333 74.6666667 76.        77.3333333 78.6666667
  80.        81.3333333 82.6666667 84.       ]]
 ```
However, if I swap 'batch_size=1' and 'sequence_length=64', then it traverses to the end of the video. 
```
[[ 0.       ]
 [ 1.3333333]
 [ 2.6666667]
 [ 4.       ]
 [ 5.3333333]
 [ 6.6666667]
 [ 8.       ]
 [ 9.3333333]
 [10.6666667]
 [12.       ]
 [13.3333333]
 [14.6666667]
 [16.       ]
 [17.3333333]
 [18.6666667]
 [20.       ]
 [21.3333333]
 [22.6666667]
 [24.       ]
 [25.3333333]
 [26.6666667]
 [28.       ]
 [29.3333333]
 [30.6666667]
 [32.       ]
 [33.3333333]
 [34.6666667]
 [36.       ]
 [37.3333333]
 [38.6666667]
 [40.       ]
 [41.3333333]
 [42.6666667]
 [44.       ]
 [45.3333333]
 [46.6666667]
 [48.       ]
 [49.3333333]
 [50.6666667]
 [52.       ]
 [53.3333333]
 [54.6666667]
 [56.       ]
 [57.3333333]
 [58.6666667]
 [60.       ]
 [61.3333333]
 [62.6666667]
 [64.       ]
 [65.3333333]
 [66.6666667]
 [68.       ]
 [69.3333333]
 [70.6666667]
 [72.       ]
 [73.3333333]
 [74.6666667]
 [76.       ]
 [77.3333333]
 [78.6666667]
 [80.       ]
 [81.3333333]
 [82.6666667]
 [84.       ]]
[[ 85.3333333]
 [ 86.6666667]
 [ 88.       ]
 [ 89.3333333]
 [ 90.6666667]
 [ 92.       ]
 [ 93.3333333]
 [ 94.6666667]
 [ 96.       ]
 [ 97.3333333]
 [ 98.6666667]
 [100.       ]
 [101.3333333]
 [102.6666667]
 [104.       ]
 [105.3333333]
 [106.6666667]
 [108.       ]
 [109.3333333]
 [110.6666667]
 [112.       ]
 [113.3333333]
 [114.6666667]
 [116.       ]
 [117.3333333]
 [118.6666667]
 [120.       ]
 [121.3333333]
 [122.6666667]
 [124.       ]
 [125.3333333]
 [126.6666667]
 [128.       ]
 [129.3333333]
 [130.6666667]
 [132.       ]
 [133.3333333]
 [134.6666667]
 [136.       ]
 [137.3333333]
 [138.6666667]
 [140.       ]
 [141.3333333]
 [142.6666667]
 [144.       ]
 [145.3333333]
 [146.6666667]
 [148.       ]
 [149.3333333]
 [150.6666667]
 [152.       ]
 [153.3333333]
 [154.6666667]
 [156.       ]
 [157.3333333]
 [158.6666667]
 [160.       ]
 [161.3333333]
 [162.6666667]
 [164.       ]
 [165.3333333]
 [166.6666667]
 [168.       ]
 [169.3333333]]
[[170.6666667]
 [172.       ]
 [173.3333333]
 [174.6666667]
 [176.       ]
 [177.3333333]
 [178.6666667]
 [180.       ]
 [181.3333333]
 [182.6666667]
 [184.       ]
 [185.3333333]
 [186.6666667]
 [  0.       ]
 [  1.3333333]
 [  2.6666667]
 [  4.       ]
 [  5.3333333]
 [  6.6666667]
 [  8.       ]
 [  9.3333333]
 [ 10.6666667]
 [ 12.       ]
 [ 13.3333333]
 [ 14.6666667]
 [ 16.       ]
 [ 17.3333333]
 [ 18.6666667]
 [ 20.       ]
 [ 21.3333333]
 [ 22.6666667]
 [ 24.       ]
 [ 25.3333333]
 [ 26.6666667]
 [ 28.       ]
 [ 29.3333333]
 [ 30.6666667]
 [ 32.       ]
 [ 33.3333333]
 [ 34.6666667]
 [ 36.       ]
 [ 37.3333333]
 [ 38.6666667]
 [ 40.       ]
 [ 41.3333333]
 [ 42.6666667]
 [ 44.       ]
 [ 45.3333333]
 [ 46.6666667]
 [ 48.       ]
 [ 49.3333333]
 [ 50.6666667]
 [ 52.       ]
 [ 53.3333333]
 [ 54.6666667]
 [ 56.       ]
 [ 57.3333333]
 [ 58.6666667]
 [ 60.       ]
 [ 61.3333333]
 [ 62.6666667]
 [ 64.       ]
 [ 65.3333333]
 [ 66.6666667]]
 ```
 
 I then apply the swap using pytorch.DALIGenericIterator and I see the following error at the last batch:
 ```
     for data in dataloader: <<<- NOTE: dataloader is class 'DALILoader' from the previous comment
  File "/<removed>/..pyenv/versions/torch/lib/python3.8/site-packages/nvidia/dali/plugin/pytorch.py", line 181, in __next__
    outputs = self._get_outputs()
  File "/<removed>/..pyenv/versions/torch/lib/python3.8/site-packages/nvidia/dali/plugin/base_iterator.py", line 246, in _get_outputs
    outputs.append(p.share_outputs())
  File "/<removed>/.pyenv/versions/torch/lib/python3.8/site-packages/nvidia/dali/pipeline.py", line 682, in share_outputs
    return self._pipe.ShareOutputs()
RuntimeError: Critical error in pipeline:
Error when executing GPU operator VideoReader encountered:
CUDA runtime API error cudaErrorInvalidValue (11):
invalid argument
Current pipeline object is no longer valid.
```
Not sure how to debug from here.

**Comment by JanuszL** [comment]
Hi,

The batch size should not matter. Only the parameters of the sequence itself - like length, stride, and step. If your sequences depend on the batch size then it is a bug and I would like to ask you for a self-contained, complete, and minimal reproduction example.

**Comment by SiftingSands** [comment]
Ok, I'll get you something by tomorrow at the latest. Just testing out ingesting multiple videos with this pipeline (hitting some snags, but that's likely out-of-scope for this ticket).

**Comment by SiftingSands** [comment]
Here's my minimal pipeline ingesting two dummy videos with different resolutions.

```
from nvidia.dali.pipeline import Pipeline
from nvidia.dali.plugin import pytorch
import nvidia.dali.fn as fn
import nvidia.dali.types as types

def create_video_reader_pipeline(batch_size, sequence_length, num_threads, 
                                 device_id, video_filenames, labels, frame_stride):
    pipeline = Pipeline(batch_size, num_threads, device_id, seed=12)
    with pipeline:
        images, label, start_frame_num, timestamps = fn.video_reader_resize(device="gpu", 
                                 filenames=video_filenames, 
                                 sequence_length=sequence_length,
                                 normalized=False, random_shuffle=False, 
                                 image_type=types.RGB, 
                                 dtype=types.UINT8, 
                                 enable_frame_num=True,
                                 enable_timestamps=True,
                                 labels=labels, 
                                 stride=frame_stride,
                                 pad_last_batch=False, name="Reader",
                                 size=[640, 640])

        pipeline.set_outputs(images, label, start_frame_num, timestamps)
    return pipeline

class DALILoader():
    def __init__(self, batch_size, video_filenames, sequence_length, frame_stride,
                 labels, num_threads, device_id):
        self.pipeline = create_video_reader_pipeline(batch_size, 
                                                     sequence_length, 
                                                     num_threads, 
                                                     device_id, 
                                                     video_filenames, 
                                                     labels,
                                                     frame_stride)
        self.pipeline.build()
        self.epoch_size = self.pipeline.epoch_size("Reader")
        
        output_names = ["data", 'label', 'start_frame_num', 'timestamps']
        self.dali_iterator = pytorch.DALIGenericIterator(self.pipeline,
                                                         output_names,
                                                         reader_name="Reader",
                                                         last_batch_policy=pytorch.LastBatchPolicy.PARTIAL,
                                                         auto_reset=False,
                                                         dynamic_shape=True) # not really needed now that we resize

    def __len__(self):
        return int(self.epoch_size)

    def __iter__(self):
        return self.dali_iterator.__iter__()


batch_size = 8
sequence_length = 1
frame_stride = 20
video_filenames = ["<INSERT_PATH>/DALI_extra/db/video/cfr_ntsc_29_97_test.mp4",
                   "/<INSERT_PATH>/DALI_extra/db/video/test_label.mp4"]
labels = list(range(len(video_filenames))) # how we keep track of sample-video pairs

dataloader = DALILoader(batch_size=batch_size, num_threads=2, device_id=0,
                        video_filenames=video_filenames, 
                        sequence_length=sequence_length,
                        frame_stride=frame_stride, labels=labels)

for data in dataloader:
    outputs = data[0]
    seq_out = outputs['data'].squeeze()

    print(seq_out.shape)
    label = outputs['label']
    start_frame_num = outputs['start_frame_num']
    timestamps = outputs['timestamps']
    print(timestamps)
    print(label)
```

This produces the following output on `nvidia-dali-cuda100    0.30.0` as I haven't updated since the last release.
I observe the expected batches with partial fills at the transition between videos.
```
torch.Size([8, 640, 640, 3])
tensor([[0.0000],
        [0.6673],
        [1.3347],
        [2.0020],
        [2.6693],
        [0.0000],
        [0.8000],
        [1.6000]], device='cuda:0', dtype=torch.float64)
tensor([[0],
        [0],
        [0],
        [0],
        [0],
        [1],
        [1],
        [1]], device='cuda:0', dtype=torch.int32)
torch.Size([8, 640, 640, 3])
tensor([[2.4000],
        [3.2000],
        [4.0000],
        [4.8000],
        [5.6000],
        [6.4000],
        [7.2000],
        [8.0000]], device='cuda:0', dtype=torch.float64)
tensor([[1],
        [1],
        [1],
        [1],
        [1],
        [1],
        [1],
        [1]], device='cuda:0', dtype=torch.int32)
torch.Size([2, 640, 640, 3])
tensor([[8.8000],
        [9.6000]], device='cuda:0', dtype=torch.float64)
tensor([[1],
        [1]], device='cuda:0', dtype=torch.int32)
```

Swapping batch_size and sequence_length gives a vastly different output. Is this the disparity that you were looking for?
```
torch.Size([8, 640, 640, 3])
tensor([[0.0000, 0.8000, 1.6000, 2.4000, 3.2000, 4.0000, 4.8000, 5.6000]],
       device='cuda:0', dtype=torch.float64)
tensor([[1]], device='cuda:0', dtype=torch.int32)
```

As an aside, not performing the resize, unsurprisingly, gives errors due to differing video resolutions. I'm assuming that there's not a built-in way to have the dataloader split the outputs by each video when it encounters a transition between videos? I don't have an immediate need for that though; just curious.

**Comment by SiftingSands** [comment]
On the subject of stride, is there a way to set it to a different number on a per video basis? For example, if we want to only capture a frame every second for processing, but we're ingesting videos that have different frame rates. I can create separate ticket for this if desired.

**Comment by JanuszL** [comment]
Hi @SiftingSands ,

> Swapping batch_size and sequence_length gives a vastly different output. Is this the disparity that you were looking for?

Swapping batch_size should not yield any difference, changing sequence_length should. Could you check only one paramater at the time as see if that yields expected result?

> As an aside, not performing the resize, unsurprisingly, gives errors due to differing video resolutions. I'm assuming that there's not a built-in way to have the dataloader split the outputs by each video when it encounters a transition between videos?

I'm sorry, but iterators expects that batches have uniform samples. MXNet and PaddlePaddle are able to support non dense tensors but in the case of PyTorch it is not possible.

> On the subject of stride, is there a way to set it to a different number on a per video basis? For example, if we want to only capture a frame every second for processing, but we're ingesting videos that have different frame rates.

This is not possible now. I think it should be doable. I will add it to our ToDo list.

**Comment by SiftingSands** [comment]
I have updated the previous comment with the specific `batch_size` and `sequence_length` used, and removed the squeeze on the tensor shape printouts. Sorry for not being clearer on that. In addition, here are my results the complement the previous findings:

```
batch_size = 8
sequence_length = 8
```
Produces:
```
torch.Size([1, 8, 640, 640, 3])
tensor([[0.0000, 0.8000, 1.6000, 2.4000, 3.2000, 4.0000, 4.8000, 5.6000]],
       device='cuda:0', dtype=torch.float64)
tensor([[1]], device='cuda:0', dtype=torch.int32)
```
And
```
batch_size = 1
sequence_length = 1
```
Produces:
```
torch.Size([1, 1, 640, 640, 3])
tensor([[0.]], device='cuda:0', dtype=torch.float64)
tensor([[0]], device='cuda:0', dtype=torch.int32)
torch.Size([1, 1, 640, 640, 3])
tensor([[0.6673]], device='cuda:0', dtype=torch.float64)
tensor([[0]], device='cuda:0', dtype=torch.int32)
torch.Size([1, 1, 640, 640, 3])
tensor([[1.3347]], device='cuda:0', dtype=torch.float64)
tensor([[0]], device='cuda:0', dtype=torch.int32)
torch.Size([1, 1, 640, 640, 3])
tensor([[2.0020]], device='cuda:0', dtype=torch.float64)
tensor([[0]], device='cuda:0', dtype=torch.int32)
torch.Size([1, 1, 640, 640, 3])
tensor([[2.6693]], device='cuda:0', dtype=torch.float64)
tensor([[0]], device='cuda:0', dtype=torch.int32)
torch.Size([1, 1, 640, 640, 3])
tensor([[0.]], device='cuda:0', dtype=torch.float64)
tensor([[1]], device='cuda:0', dtype=torch.int32)
torch.Size([1, 1, 640, 640, 3])
tensor([[0.8000]], device='cuda:0', dtype=torch.float64)
tensor([[1]], device='cuda:0', dtype=torch.int32)
torch.Size([1, 1, 640, 640, 3])
tensor([[1.6000]], device='cuda:0', dtype=torch.float64)
tensor([[1]], device='cuda:0', dtype=torch.int32)
torch.Size([1, 1, 640, 640, 3])
tensor([[2.4000]], device='cuda:0', dtype=torch.float64)
tensor([[1]], device='cuda:0', dtype=torch.int32)
torch.Size([1, 1, 640, 640, 3])
tensor([[3.2000]], device='cuda:0', dtype=torch.float64)
tensor([[1]], device='cuda:0', dtype=torch.int32)
torch.Size([1, 1, 640, 640, 3])
tensor([[4.]], device='cuda:0', dtype=torch.float64)
tensor([[1]], device='cuda:0', dtype=torch.int32)
torch.Size([1, 1, 640, 640, 3])
tensor([[4.8000]], device='cuda:0', dtype=torch.float64)
tensor([[1]], device='cuda:0', dtype=torch.int32)
torch.Size([1, 1, 640, 640, 3])
tensor([[5.6000]], device='cuda:0', dtype=torch.float64)
tensor([[1]], device='cuda:0', dtype=torch.int32)
torch.Size([1, 1, 640, 640, 3])
tensor([[6.4000]], device='cuda:0', dtype=torch.float64)
tensor([[1]], device='cuda:0', dtype=torch.int32)
torch.Size([1, 1, 640, 640, 3])
tensor([[7.2000]], device='cuda:0', dtype=torch.float64)
tensor([[1]], device='cuda:0', dtype=torch.int32)
torch.Size([1, 1, 640, 640, 3])
tensor([[8.]], device='cuda:0', dtype=torch.float64)
tensor([[1]], device='cuda:0', dtype=torch.int32)
torch.Size([1, 1, 640, 640, 3])
tensor([[8.8000]], device='cuda:0', dtype=torch.float64)
tensor([[1]], device='cuda:0', dtype=torch.int32)
torch.Size([1, 1, 640, 640, 3])
tensor([[9.6000]], device='cuda:0', dtype=torch.float64)
tensor([[1]], device='cuda:0', dtype=torch.int32)

```

**Comment by JanuszL** [comment]
That is expected. Consider the video with frames with the following timestamps [0, 1, 3, 4, 5].
If you set sequence length to 1 you can expect outputs:
- 0
- 1
- 2
- 3
- 4
- 5
If you set it to 4
- 0
- 1
As there is no sequence of length 4 starting frames 2, 3, 4, and 5. The batch size should not have anything to do with this.

**Comment by SiftingSands** [comment]
Got it, that makes sense. If everything checks, out I'll close this unless you want to keep it open for the variable stride?

**Comment by JanuszL** [comment]
Let keep it open if you want to track this request in this issue. Or create a separate one for this particular future request.

**Comment by SiftingSands** [comment]
I'll create a separate issue to not overload this one.

**Comment by SiftingSands** [comment]
@JanuszL Sorry to resurrect this old issue. When I fix `sequence_length=1` to capture the ends of a video (otherwise there won't be a valid sequence as you mentioned), GPU utilization is very low. I haven't timed it, but the difference is very noticeable, e.g. `sequence_length=1; batch_size=32` is much slower than `sequence_length=32, batch_size=1`. I'm assuming this is due to far more I/O when the sequence_length is small? I've been trying to figure out how to keep things fast while still being able to read in the last parts of a video that don't fit into a batch with no avail. Any suggestions?

**Comment by JanuszL** [comment]
Hi,

You should check the decoder utilization:
```nvidia-smi -q -g 0 -d UTILIZATION -l```.
DALI's video reader decodes sequences from the video. The flow is:
- seek to the first key frame that precedes the first frame from the sequence
- decode consecutive frames till the last one from the sequence

So the shorter the sequence is and/or the bigger step between frames in the sequence the more the decoder needs to decode, and more frames are dropped.
I don't think we can do much about it now. The solution we are thinking about are:
- pad the missing frames in the last sequence
- support variable-length sequences (although it is a bigger effort to adjust operators)

**Comment by JanuszL** [comment]
We will add it to our ToDo list, but I cannot tell when we get to it. If you want you can try to implement this on your own and create a PR. We would be more than happy to review it.

**Comment by SiftingSands** [comment]
Understood. Could you suggest on a good place to start? For example, is it sufficient to modify something like 'fn.readers.video', or would one need to go into a lower abstraction (admittedly my C/++ skills are lackluster).

**Comment by JanuszL** [comment]
@SiftingSands,

I would start with:
- https://github.com/NVIDIA/DALI/blob/master/dali/operators/reader/loader/video_loader.h#L267 - this creates a list of all available sequences. You need to change it in the way that a last, shorter one is allowed too. Also, you need to indicate that this sequence is shorter (additional field in `sequence_meta` maybe)
- https://github.com/NVIDIA/DALI/blob/master/dali/operators/reader/loader/video_loader.cc#L732 need to be adjusted to read less than `count_` frames, and then you can optionally launch cudaMemset to zero out the remaining frames to avoid having garbage there
- add a test in https://github.com/NVIDIA/DALI/blob/master/dali/test/python/test_video_pipeline.py to make sure that padding works as expected

**Comment by SiftingSands** [comment]
Thanks for the advice! I've been poking around the source code, and I've had some success on getting the last few frames of a video. (Doing some jank slicing of the output timestamps on the Python side though) However, I'm getting hangs on run-time when using strides > 1 when it tries to seek the last frame. In that train of thought, could elaborate on why `total_count = 1 + (count - 1) * stride_;` in `video_loader.cc and video_loader.h` is not just `count * stride`? Is there some off-by-1 logic that it is accounting for?

**Comment by JanuszL** [comment]
Hi @SiftingSands,
>  In that train of thought, could elaborate on why total_count = 1 + (count - 1) * stride_; in video_loader.cc and video_loader.h is not just count * stride? Is there some off-by-1 logic that it is accounting for

You need X frames, there are X-1 strides in between. Like, count=3, stride=2 `frame-stride-frame-stride-frame` and the number of frames to read is as staged above - 5 (frame-2-frame-2-frame).

**Comment by SiftingSands** [comment]
Thanks for the explanation. Yeah, that should have been clear to me in hindsight. I'm gonna blame this on reading too many debugging messages haha.

**Comment by SiftingSands** [comment]
@JanuszL 
I've got the aforementioned functionality working by resizing the tensor on the last sequence of the video. However, I realize that won't work for a batch_size > 1, but that is sufficient for my current application. 

I attempted to create a padded sequence to handle the `batch_size>1` case, without success. For example, I keep the tensor initialization unchanged in `video_loader.cc` and `VideoLoader::ReadSample` :
`tensor.initialize(count_, seq_meta.height, seq_meta.width, channels_, dtype_);`
and then change the number of frames to be read based on another `int` that was added to the `sequence_meta` struct in `video_loader.h`:
```
    tensor.read_sample_f = [this,
                            file_name = file_info_[seq_meta.filename_idx].video_file,
                            index = seq_meta.frame_idx, count = seq_meta.seq_count, &tensor] () {
      push_sequence_to_read(file_name, index, count);
      receive_frames(tensor);
    };
```
This now hangs at the end of the video, because I believe it's still waiting to fill up `tensor` to `count_` instead of `seq_meta.seq_count`. 

For zeroing out the sequence, I tried `CUDA_CALL(cudaMemset(&tensor.sequence, 0, tensor.sequence.num_bytes_));` also within `VideoLoader::ReadSample`. It's definitely incorrect, but I'm not sure if it's solely because `tensor.sequence.num_bytes_` is protected, or if it needs to be done before or within `tensor.read_sample_f`.

Finally, for the associated timestamp outputs; my first thought was to initialize `timestamp_output_` in `video_reader_op.h` to `-1` to indicate that it corresponds to a padded frame. However, I'm not sure if that's compatible with other parts of DALI, or if that can be accomplished within `video_reader_op.h` or the `GetSample` method has to be modified `pipeline/workspace/host_workspace.cc`.

I would appreciate any suggestions, but I might have to leave this here for now due to other priorities (depending on if I can resolve these issues quickly).
